src/Substrait.Core/ProtoConverter.cs

```diff
@@ -0,0 +1,154 @@
+        ProtoRel.RelTypeOneofCase.Sort => CreateSort(protoRel.Sort),
+        _ => throw new NotImplementedException(relType.ToString()),
+      };
+    }
+
+    /// <summary>
+    /// Create domain object specific to relational expression type from protobuf message
+    /// </summary>
+    /// <param name="protoRex">Relational expression encoded as protobuf message</param>
+    /// <returns><see cref="Rel"/> corresponding to message's expression type</returns>
+    private RelExpression ToExpression(ProtoRex protoRex)
+    {
+      var rexType = protoRex.RexTypeCase;
+
+      return rexType switch
+      {
+        ProtoRex.RexTypeOneofCase.Selection => CreateFieldReference(protoRex.Selection),
+        ProtoRex.RexTypeOneofCase.None => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Literal => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.ScalarFunction => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.WindowFunction => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.IfThen => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.SwitchExpression => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.SingularOrList => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.MultiOrList => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Cast => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Subquery => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Nested => throw new NotImplementedException(),
+        ProtoRex.RexTypeOneofCase.Enum => throw new NotImplementedException(),
+        _ => throw new NotImplementedException(rexType.ToString()),
+      };
+    }
+
+    private Rel CreateAgg(AggregateRel agg)
+    {
+      var input = ToRel(agg.Input);
+      var inputs = ImmutableList.Create(input);
+      return new Aggregate()
+      {
+        Inputs = inputs,
+      };
+    }
+
+    private Rel CreateCross(CrossRel cross)
+    {
+      var left = ToRel(cross.Left);
+      var right = ToRel(cross.Right);
+      var inputs = ImmutableList.Create(left, right);
+      return new Cross()
+      {
+        Inputs = inputs,
+      };
+    }
+
+    private Rel CreateFetch(FetchRel fetch)
+    {
+      var input = ToRel(fetch.Input);
+      var inputs = ImmutableList.Create(input);
+      return new Fetch()
+      {
+        Count = fetch.Count,
+        Inputs = inputs,
+      };
+    }
+
+    private Rel CreateFilter(FilterRel filter)
+    {
+      var input = ToRel(filter.Input);
+      var inputs = ImmutableList.Create(input);
+      return new Filter()
+      {
+        Inputs = inputs,
+      };
+    }
+
+    private Rel CreateJoin(JoinRel join)
+    {
+      var left = ToRel(join.Left);
+      var right = ToRel(join.Right);
+      var _ = ToExpression(join.Expression);
```

Comment:
This seems a touch arbitrary (e.g. why grab join expression and not the join filter?  why not grab a filter's filter expression?) but there's probably no reason to get hung up on it.